File: Queue/LinearQueue/LinearQueue.py
class LinearQueue:
    def __init__(self, capacity=5):
        self.capacity = capacity
        self.queue = [None] * capacity
        self.front_idx = 0
        self.rear_idx = -1
        self.current_size = 0

    def isFull(self):
        return self.current_size == self.capacity

    def isEmpty(self):
        return self.current_size == 0

    def enqueue(self, value):
        if self.isFull():
            print(f"Queue Overflow! Cannot enqueue {value}")
            return False
        self.rear_idx += 1
        self.queue[self.rear_idx] = value
        self.current_size += 1
        return True

    def dequeue(self):
        if self.isEmpty():
            print("Queue Underflow! Cannot dequeue")
            return None
        value = self.queue[self.front_idx]
        self.queue[self.front_idx] = None  
        self.front_idx += 1
        self.current_size -= 1
        return value

    def peek(self):
        if self.isEmpty():
            print("Queue is empty!")
            return None
        return self.queue[self.front_idx]

    def size(self):
        return self.current_size

    def print(self):
        if self.isEmpty():
            print("Queue is empty")
            return
        print("Queue (Front -> Rear): ", end="")
        elements = [str(self.queue[i]) for i in range(self.front_idx, self.rear_idx + 1)]
        print(" -> ".join(elements))

if __name__ == "__main__":
    q = LinearQueue(3)
    q.enqueue(10)
    q.enqueue(20)
    q.enqueue(30)
    q.enqueue(40)  # Overflow
    q.print()

    print("Dequeued:", q.dequeue())
    print("Front element:", q.peek())
    q.print()

File: Queue/LinearQueue/LinearQueue.java
public class LinearQueue {
    private int[] data;
    private int front;
    private int rear;
    private int size;
    private int capacity;

    public LinearQueue(int capacity) {
        this.capacity = capacity;
        this.data = new int[capacity];
        this.front = 0;
        this.rear = -1;
        this.size = 0;
    }

    public boolean isFull() {
        return size == capacity;
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public boolean enqueue(int value) {
        if (isFull()) {
            System.out.println("Queue Overflow! Cannot enqueue " + value);
            return false;
        }
        data[++rear] = value;
        size++;
        return true;
    }

    public int dequeue() {
        if (isEmpty()) {
            System.out.println("Queue Underflow! Cannot dequeue");
            return -1;
        }
        size--;
        return data[front++];
    }

    public int peek() {
        if (isEmpty()) {
            System.out.println("Queue is empty!");
            return -1;
        }
        return data[front];
    }

    public int size() {
        return size;
    }

    public void print() {
        if (isEmpty()) {
            System.out.println("Queue is empty");
            return;
        }
        System.out.print("Queue (Front -> Rear): ");
        for (int i = front; i <= rear; i++) {
            System.out.print(data[i] + " ");
        }
        System.out.println();
    }

    public static void main(String[] args) {
        LinearQueue q = new LinearQueue(5);
        q.enqueue(10);
        q.enqueue(20);
        q.enqueue(30);
        q.print();

        System.out.println("Dequeued: " + q.dequeue());
        System.out.println("Front element: " + q.peek());
        q.print();
    }
}
